src/Swift.Bindings/src/Model/TypeSpecParsing/TypeSpecParser.cs

```diff
@@ -0,0 +1,358 @@
+/// It will lool like:
+/// typespec: [attribute]*[inout][any][typelable]typespec[generic-clause]['?']
+/// typespec: tuple-typespec|named-typespec|func-typespec|protocol-list-type-spec
+/// tuple-typespec:'('typespec-list')'
+/// named-typespec: name
+/// func-type-spec: ypespec [async][throws]t '->' typespec
+/// protocol-list-type-spec: named-typespec &amp; named-typespec [&amp; named-type-spec]*
+/// generic-clause: '<' typespec-list '>'
+/// typespeclist: typespec [',' typespec]*
+/// attribute @name['(' parameters ')']
+/// </summary>
+public class TypeSpecParser {
+    TextReader reader;
+    TypeSpecTokenizer tokenizer;
+
+    /// <summary>
+    /// Private constructor for a TypeSpecParser using the supplied TextReader as a source
+    /// </summary>
+    TypeSpecParser(TextReader reader)
+    {
+        this.reader = reader;
+        tokenizer = new TypeSpecTokenizer(reader);
+    }
+
+    /// <summary>
+    /// Private utility method to parse and return a TypeSpec
+    /// </summary>
+    TypeSpec? Parse()
+    {
+        // This is a very simple recursive descent parser with 1 token look-ahead
+        // The typespec language is very simple and doesn't need anything more
+        var token = tokenizer.Peek();
+        TypeSpec? type = null;
+        List<TypeSpecAttribute> attrs = new ();
+        var inout = false;
+        var isAny = false;
+        string? typeLabel = null;
+        var throwsClosure = false;
+        var asyncClosure = false;
+        var expectClosure = false;
+
+        // Prefix
+
+        // parse any attributes
+        if (token.Kind == TypeTokenKind.At) {
+            attrs = ParseAttributes();
+            token = tokenizer.Peek();
+        }
+
+        // looks like it's inout, consume and continue
+        if (token.Kind == TypeTokenKind.TypeName && token.Value == "inout") {
+            inout = true;
+            tokenizer.Next ();
+            token = tokenizer.Peek ();
+        }
+
+        // any, consume and continute
+        if (token.Kind == TypeTokenKind.TypeName && token.Value == "any") {
+            isAny = true;
+            tokenizer.Next ();
+            token = tokenizer.Peek ();
+        }
+
+        // label, consume and continue
+        if (token.Kind == TypeTokenKind.TypeLabel) {
+            typeLabel = token.Value;
+            tokenizer.Next ();
+            token = tokenizer.Peek ();
+        }
+
+        // meat of the type
+
+        if (token.Kind == TypeTokenKind.LeftParenthesis) { // tuple
+            tokenizer.Next ();
+            TupleTypeSpec tuple = ParseTuple ();
+            type = tuple.Elements.Count == 1 ? tuple.Elements [0] : tuple;
+            typeLabel = type.TypeLabel;
+            type.TypeLabel = null;
+        } else if (token.Kind == TypeTokenKind.TypeName) { // name
+            tokenizer.Next ();
+            var tokenValue = token.Value.StartsWith ("ObjectiveC.", StringComparison.Ordinal) ?
```

Comment:
In this case, this was a bug/feature from the Swift compiler in that it was emitting ObjC types that should have been Foundation types.

Reply:
Interesting, I haven't known about this. Does it mean that e.g., instead of `Foundation.NSString` it would be `ObjectiveC.NSString`, right? 

You mentioned it is a bug/feature in Swift compiler. Do you know if it is still valid and we need this?